cpp/src/groupby/groupby.cu

```diff
@@ -138,6 +230,57 @@ detail::sort::sort_groupby_helper& groupby::helper() {
   return *_helper;
 };
 
+std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> groupby::windowed_aggregate(
+    std::vector<window_aggregation_request> const& requests,
+    rmm::mr::device_memory_resource* mr) {
+
+  CUDF_EXPECTS(std::all_of(requests.begin(), requests.end(),
+                           [this](auto const& request) {
+                             return request.values.size() == _keys.num_rows();
+                           }),
+               "Size mismatch between request values and groupby keys.");
+
+  CUDF_EXPECTS(this->_keys_are_sorted, 
+               "Window-aggregation is currently supported only on pre-sorted key columns.");
+
+  if (_keys.num_rows() == 0) {
+    std::make_pair(empty_like(_keys), 
+                  templated_empty_results(requests, 
+                                          [](std::pair<window_bounds, std::unique_ptr<aggregation>> const& agg) 
+                                          {return agg.second->kind;}));
+  }
+
+  auto group_offsets = helper().group_offsets();
+  group_offsets.push_back(_keys.num_rows()); // Cap the end.
```

Comment:
@devavret  `helper::group_offsets` should probably be modified to return `n+1` offsets to avoid needing to do this.